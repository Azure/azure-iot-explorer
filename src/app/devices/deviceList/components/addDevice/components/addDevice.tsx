/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { RouteComponentProps, Route } from 'react-router-dom';
import { Overlay } from 'office-ui-fabric-react/lib/Overlay';
import { Toggle } from 'office-ui-fabric-react/lib/Toggle';
import { ChoiceGroup, IChoiceGroupOption } from 'office-ui-fabric-react/lib/ChoiceGroup';
import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox';
import { CommandBar } from 'office-ui-fabric-react/lib/CommandBar';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../../../localization/resourceKeys';
import { DeviceAuthenticationType } from '../../../../../api/models/deviceAuthenticationType';
import { DeviceIdentity } from '../../../../../api/models/deviceIdentity';
import { DeviceStatus } from '../../../../../api/models/deviceStatus';
import { validateKey, validateThumbprint, validateDeviceId } from '../../../../../shared/utils/utils';
import LabelWithTooltip from '../../../../../shared/components/labelWithTooltip';
import MaskedCopyableTextFieldContainer from '../../../../../shared/components/maskedCopyableTextFieldContainer';
import { SynchronizationStatus } from '../../../../../api/models/synchronizationStatus';
import { SAVE, CANCEL } from '../../../../../constants/iconNames';
import Breadcrumb from '../../../../../shared/components/breadcrumb';
import '../../../../../css/_addDevice.scss';
import '../../../../../css/_layouts.scss';

export interface AddDeviceActionProps {
    handleSave: (deviceIdentity: DeviceIdentity) => void;
}

export interface AddDeviceDataProps {
    deviceListSyncStatus: SynchronizationStatus;
}

export interface AddDeviceState {
    authenticationType: DeviceAuthenticationType;
    autoGenerateKeys: boolean;
    primaryKey: string;
    secondaryKey: string;
    primaryThumbprint: string;
    secondaryThumbprint: string;
    deviceId: string;
    status: DeviceStatus;

    deviceIdError?: string;
    primaryKeyError?: string;
    secondaryKeyError?: string;
    primaryThumbprintError?: string;
    secondaryThumbprintError?: string;
}

export default class AddDevice extends React.Component<AddDeviceActionProps & AddDeviceDataProps & RouteComponentProps, AddDeviceState> {

    constructor(props: AddDeviceActionProps & AddDeviceDataProps & RouteComponentProps) {
        super(props);

        this.state = {
            authenticationType: DeviceAuthenticationType.SymmetricKey,
            autoGenerateKeys: true,
            deviceId: '',
            primaryKey: '',
            primaryThumbprint: '',
            secondaryKey: '',
            secondaryThumbprint: '',
            status: DeviceStatus.Enabled
        };
    }

    public render() {
        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                     <form onSubmit={this.handleSave} className="view add-device">
                        <div className="view-command">
                            {this.showCommandBar(context)}
                        </div>
                        <div className="view-content view-scroll-vertical">
                            <div className="form">
                                {this.showDeviceId(context)}
                                {this.showAuthentication(context)}
                                {this.showConnectivity(context)}
                            </div>
                        </div>
                        {this.props.deviceListSyncStatus === SynchronizationStatus.updating && <Overlay/>}
                    </form>
                )}
            </LocalizationContextConsumer>
        );
    }

    public componentDidUpdate() {
        if (this.props.deviceListSyncStatus === SynchronizationStatus.upserted) { // only when device has been added successfully would navigate to list view
            this.navigateToDeviceList();
        }
    }

    private readonly navigateToDeviceList = () => {
        const path = this.props.match.url.replace(/\/add/, ``);
        this.props.history.push(path);
    }

    private readonly showDeviceId = (context: LocalizationContextInterface) => {
        return (
            <MaskedCopyableTextFieldContainer
                ariaLabel={context.t(ResourceKeys.deviceIdentity.deviceID)}
                label={context.t(ResourceKeys.deviceIdentity.deviceID)}
                value={this.state.deviceId}
                required={true}
                onTextChange={this.changeDeviceId}
                allowMask={false}
                readOnly={false}
                error={!!this.state.deviceIdError ? context.t(this.state.deviceIdError) : ''}
                labelCallout={context.t(ResourceKeys.deviceIdentity.deviceIDTooltip)}
                setFocus={true}
            />
        );
    }

    private readonly getAuthType = (context: LocalizationContextInterface) => {
        return (
            <ChoiceGroup
                label={context.t(ResourceKeys.deviceIdentity.authenticationType.text)}
                selectedKey={this.state.authenticationType}
                onChange={this.changeAuthenticationType}
                options={
                    [
                        {
                            key: DeviceAuthenticationType.SymmetricKey,
                            text: context.t(ResourceKeys.deviceIdentity.authenticationType.symmetricKey.type)
                        },
                        {
                            key: DeviceAuthenticationType.SelfSigned,
                            text: context.t(ResourceKeys.deviceIdentity.authenticationType.selfSigned.type)
                        },
                        {
                            key: DeviceAuthenticationType.CACertificate,
                            text: context.t(ResourceKeys.deviceIdentity.authenticationType.ca.type)
                        }
                    ]
                }
                required={true}
            />
        );
    }

    private readonly renderSymmetricKeySection = (context: LocalizationContextInterface) => {
        return (
            <>
                <MaskedCopyableTextFieldContainer
                    ariaLabel={context.t(ResourceKeys.deviceIdentity.authenticationType.symmetricKey.primaryKey)}
                    label={context.t(ResourceKeys.deviceIdentity.authenticationType.symmetricKey.primaryKey)}
                    value={this.state.primaryKey}
                    required={true}
                    onTextChange={this.changePrimaryKey}
                    allowMask={true}
                    readOnly={false}
                    error={!!this.state.primaryKeyError ? context.t(this.state.primaryKeyError) : ''}
                    labelCallout={context.t(ResourceKeys.deviceIdentity.authenticationType.symmetricKey.primaryKeyTooltip)}
                />
                <MaskedCopyableTextFieldContainer
                    ariaLabel={context.t(ResourceKeys.deviceIdentity.authenticationType.symmetricKey.secondaryKey)}
                    label={context.t(ResourceKeys.deviceIdentity.authenticationType.symmetricKey.secondaryKey)}
                    value={this.state.secondaryKey}
                    required={true}
                    onTextChange={this.changeSecondaryKey}
                    allowMask={true}
                    readOnly={false}
                    error={!!this.state.secondaryKeyError ? context.t(this.state.secondaryKeyError) : ''}
                    labelCallout={context.t(ResourceKeys.deviceIdentity.authenticationType.symmetricKey.secondaryKeyTooltip)}
                />
            </>
        );
    }

    private readonly renderSelfSignedSection = (context: LocalizationContextInterface) => {
        return (
            <>
                <MaskedCopyableTextFieldContainer
                    ariaLabel={context.t(ResourceKeys.deviceIdentity.authenticationType.selfSigned.primaryThumbprint)}
                    label={context.t(ResourceKeys.deviceIdentity.authenticationType.selfSigned.primaryThumbprint)}
                    value={this.state.primaryThumbprint}
                    required={true}
                    onTextChange={this.changePrimaryThumbprint}
                    allowMask={true}
                    readOnly={false}
                    error={!!this.state.primaryThumbprintError ? context.t(this.state.primaryThumbprintError) : ''}
                    labelCallout={context.t(ResourceKeys.deviceIdentity.authenticationType.selfSigned.primaryThumbprintTooltip)}
                />
                <MaskedCopyableTextFieldContainer
                    ariaLabel={context.t(ResourceKeys.deviceIdentity.authenticationType.selfSigned.secondaryThumbprint)}
                    label={context.t(ResourceKeys.deviceIdentity.authenticationType.selfSigned.secondaryThumbprint)}
                    value={this.state.secondaryThumbprint}
                    required={true}
                    onTextChange={this.changeSecondaryThumbprint}
                    allowMask={true}
                    readOnly={false}
                    error={!!this.state.secondaryThumbprintError ? context.t(this.state.secondaryThumbprintError) : ''}
                    labelCallout={context.t(ResourceKeys.deviceIdentity.authenticationType.selfSigned.secondaryThumbprintTooltip)}
                />
            </>
        );
    }

    private readonly showAuthentication = (context: LocalizationContextInterface) => {
        return (
            <div className="authentication">
                {this.getAuthType(context)}
                {this.state.authenticationType === DeviceAuthenticationType.SymmetricKey &&
                    <>
                        <Checkbox
                            className="autoGenerateButton"
                            label={context.t(ResourceKeys.deviceIdentity.authenticationType.symmetricKey.autoGenerate)}
                            checked={this.state.autoGenerateKeys}
                            onChange={this.changeAutoGenerateKeys}
                        />
                        {!this.state.autoGenerateKeys &&
                            this.renderSymmetricKeySection(context)
                        }
                    </>
                }
                {this.state.authenticationType === DeviceAuthenticationType.SelfSigned &&
                    this.renderSelfSignedSection(context)
                }
            </div>
        );
    }

    private readonly showConnectivity = (context: LocalizationContextInterface) => {

        return (
            <div className="connectivity">
                <LabelWithTooltip
                    tooltipText={context.t(ResourceKeys.deviceIdentity.hubConnectivity.tooltip)}
                >
                    {context.t(ResourceKeys.deviceIdentity.hubConnectivity.label)}
                </LabelWithTooltip>
                <Toggle
                    ariaLabel={context.t(ResourceKeys.deviceIdentity.hubConnectivity.label)}
                    checked={this.state.status === DeviceStatus.Enabled}
                    onText={context.t(ResourceKeys.deviceIdentity.hubConnectivity.enable)}
                    offText={context.t(ResourceKeys.deviceIdentity.hubConnectivity.disable)}
                    onChange={this.changeDeviceStatus}
                />
            </div>
        );
    }

    private readonly showCommandBar = (context: LocalizationContextInterface) => {
        return (
            <CommandBar
                items={[
                    {
                        ariaLabel: context.t(ResourceKeys.deviceLists.commands.save),
                        disabled: this.disableSaveButton(),
                        iconProps: {
                            iconName: SAVE
                        },
                        key: SAVE,
                        name: context.t(ResourceKeys.deviceLists.commands.save),
                        type: 'submit'
                    },
                    {
                        ariaLabel: context.t(ResourceKeys.deviceLists.commands.close),
                        disabled: false,
                        iconProps: {
                            iconName: CANCEL
                        },
                        key: CANCEL,
                        name: context.t(ResourceKeys.deviceLists.commands.close),
                        onClick: this.navigateToDeviceList
                    },
                ]}
            />
        );
    }

    private readonly changeDeviceId = (newValue?: string) => {
        const deviceIdError = this.getDeviceIdValidationMessage(newValue);
        this.setState({
            deviceId: newValue,
            deviceIdError
        });
    }

    private readonly changeAuthenticationType = (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, option?: IChoiceGroupOption) => {
        this.setState({
            authenticationType: option.key as DeviceAuthenticationType
        });
    }

    private readonly changeDeviceStatus = (event: React.MouseEvent<HTMLElement>, checked?: boolean) => {
        this.setState(
            {
                status: checked ? DeviceStatus.Enabled : DeviceStatus.Disabled
            }
        );
    }

    private readonly changeAutoGenerateKeys = (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, checked?: boolean) => {
        this.setState(
            {
                autoGenerateKeys: checked
            }
        );
        if (checked) {
            this.setState(
                {
                    primaryKey: undefined,
                    secondaryKey: undefined
                }
            );
        }
    }

    private readonly changePrimaryKey = (newValue?: string) => {
        const primaryKeyError = this.getSymmetricKeyValidationMessage(newValue);
        this.setState({
            primaryKey: newValue,
            primaryKeyError
        });
    }

    private readonly changeSecondaryKey = (newValue?: string) => {
        const secondaryKeyError = this.getSymmetricKeyValidationMessage(newValue);
        this.setState({
            secondaryKey: newValue,
            secondaryKeyError
        });
    }

    private readonly changePrimaryThumbprint = (newValue?: string) => {
        const primaryThumbprintError = this.getThumbprintValidationMessage(newValue);
        this.setState({
            primaryThumbprint: newValue,
            primaryThumbprintError
        });
    }

    private readonly changeSecondaryThumbprint = (newValue?: string) => {
        const secondaryThumbprintError = this.getThumbprintValidationMessage(newValue);
        this.setState({
            secondaryThumbprint: newValue,
            secondaryThumbprintError
        });
    }

    // tslint:disable-next-line:cyclomatic-complexity
    private readonly disableSaveButton = () => {
        if (!this.state.deviceId || !validateDeviceId(this.state.deviceId)) { return true; }
        if (this.state.authenticationType === DeviceAuthenticationType.SymmetricKey) {
            if (!this.state.autoGenerateKeys) {
                return ! (this.state.primaryKey &&
                this.state.secondaryKey &&
                validateKey(this.state.primaryKey) &&
                validateKey(this.state.secondaryKey));
            }
        }
        if (this.state.authenticationType === DeviceAuthenticationType.SelfSigned) {
            return ! (this.state.primaryThumbprint &&
            this.state.secondaryThumbprint &&
            validateThumbprint(this.state.primaryThumbprint) &&
            validateThumbprint(this.state.secondaryThumbprint));
        }
        return false;
    }

    private readonly getDeviceIdentity = (): DeviceIdentity => {
        return {
            authentication: {
                symmetricKey: this.state.authenticationType === DeviceAuthenticationType.SymmetricKey ? {
                    primaryKey: this.state.autoGenerateKeys ? null : this.state.primaryKey,
                    secondaryKey: this.state.autoGenerateKeys ? null : this.state.secondaryKey
                } : null,
                type: this.state.authenticationType.toString(),
                x509Thumbprint: this.state.authenticationType === DeviceAuthenticationType.SelfSigned ? {
                    primaryThumbprint: this.state.primaryThumbprint,
                    secondaryThumbprint: this.state.secondaryThumbprint
                } : null,
            },
            capabilities: {
                iotEdge: false
            },
            cloudToDeviceMessageCount: null,
            deviceId: this.state.deviceId,
            etag: null,
            lastActivityTime: null,
            status: this.state.status.toString(),
            statusReason: null,
            statusUpdatedTime: null
        };
    }

    private getDeviceIdValidationMessage = (value: string) => {
        return validateDeviceId(value) ? '' : ResourceKeys.deviceIdentity.validation.invalidDeviceId;
    }

    private getSymmetricKeyValidationMessage = (value: string): string => {
        return this.state.autoGenerateKeys ? '' : validateKey(value) ? '' : ResourceKeys.deviceIdentity.validation.invalidKey;
    }

    private getThumbprintValidationMessage = (value: string) => {
        return validateThumbprint(value) ? '' : ResourceKeys.deviceIdentity.validation.invalidThumbprint;
    }

    private readonly handleSave = (event: React.FormEvent<HTMLFormElement>) => {
        // Prevent page refresh
        event.preventDefault();
        this.props.handleSave(this.getDeviceIdentity());
    }
}
