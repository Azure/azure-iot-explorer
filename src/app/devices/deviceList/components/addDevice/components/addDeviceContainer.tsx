/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import AddDevice, { AddDeviceActionProps, AddDeviceDataProps } from './addDevice';
import { DeviceIdentity } from '../../../../../api/models/deviceIdentity';
import { addDeviceAction, listDevicesAction } from '../../../actions';
import { StateType } from '../../../../../shared/redux/state';
import { getDeviceSummaryListStatus } from '../../../selectors';

const mapDispatchToProps = (dispatch: Dispatch): AddDeviceActionProps => {
    return {
        handleSave: (deviceIdentity: DeviceIdentity) => dispatch(addDeviceAction.started(deviceIdentity)),
    };
};

const mapStateToProps = (state: StateType): AddDeviceDataProps => {
    return {
        deviceListSyncStatus: getDeviceSummaryListStatus(state)
    };
};

export default connect(mapStateToProps, mapDispatchToProps, undefined, {
    pure: false,
})(AddDevice);
