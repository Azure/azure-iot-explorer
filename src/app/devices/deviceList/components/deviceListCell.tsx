import * as React from 'react';
import { Icon, registerIcons } from 'office-ui-fabric-react';
import { DeviceSummary } from '../../../api/models/deviceSummary';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import '../../../css/_deviceListCell.scss';

// tslint:disable
registerIcons({
    icons: {
      'pnp-svg': (
        <svg version="1.1" xmlns="http://www.w3.org/2000/svg" x="0px" y="0px" viewBox="0 0 64 64" enable-background="new 0 0 64 64">
        <path fill="#2F76BC" d="M58.7,9.9L58.7,9.9c-0.4,0-0.8,0-0.8,0.4L50.2,20c-2.8-6.9-9.3-11.7-17.4-11.3c-8.1,0-15,4.9-17.8,12.2
            C6.9,21.6,0,27.7,0,37.4S8.5,54,17.8,54h31.6C57.5,54,64,47.1,64,39.8c0-6.1-4.1-11.3-9.7-13l9.3-12.2c0,0,0.4-0.4,0-0.8l0,0V9.9
            L58.7,9.9z"/>
        <path fill="#75D0EB" d="M17.8,32.9l-1.2-1.6c-0.4-0.4-0.4-1.2,0-1.6l4.1-3.6c0.4,0,0.4-0.4,0.8-0.4c0.4,0,0.4,0,0.8,0.4l10.9,11.7
            l24.7-32c0.4-0.4,0.4-0.4,0.8-0.4c0.4,0,0.4,0,0.4,0l4.5,3.2C64,8.6,64,9,64,9.5c0,0.4,0,0.4-0.4,0.8L34,48.3
            c-0.4,0.4-1.2,0.4-1.6,0L17.8,32.9z"/>
       </svg>
      )
    }
  });
//tslint: enable

export interface DeviceListCellProps {
    itemIndex: number | string;
    device: DeviceSummary;
    deviceId: string;
    isLoading: boolean;
    interfaceIds: string[];
    fetchDeviceInfo: (deviceId: string) => void;
}

export class DeviceListCell extends React.PureComponent<DeviceListCellProps> {

    constructor(props: DeviceListCellProps) {
        super(props);
    }

    public render() {
        const { itemIndex, isLoading, device } = this.props;

        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                    !isLoading ?
                        <div className="device-list-cell-container" data-selection-index={itemIndex}>
                            {this.renderCellDeviceInfo(device, context)}
                            {this.renderCellInterfaceInfo(device, context)}
                        </div>
                    :
                        <div className="device-list-cell-container" data-selection-index={itemIndex}>
                            {this.renderLoadingInfo(context)}
                        </div>
                )}
            </LocalizationContextConsumer>
        );
    }

    public componentWillMount() {
        const { fetchDeviceInfo, device } = this.props;
        fetchDeviceInfo(device.deviceId);
    }

    private readonly renderCellDeviceInfo = (device: DeviceSummary, context: LocalizationContextInterface) => {
        return (
            <div className="device-list-cell-container-content">
                <span className="device-list-cell-item first">{`${context.t(ResourceKeys.deviceLists.columns.lastActivityTime)}: `}<span className="data">{device.lastActivityTime || context.t(ResourceKeys.deviceLists.noData)}</span></span>
                <span className={`device-list-cell-item ${!device.isPnpDevice && 'last'}`}>{`${context.t(ResourceKeys.deviceLists.columns.statusUpdatedTime)}: `}<span className="data">{device.statusUpdatedTime || context.t(ResourceKeys.deviceLists.noData)}</span></span>
                {!!device.isPnpDevice &&
                    <span className="device-list-cell-item last">
                        <Icon
                            iconName="pnp-svg"
                            className="icon"
                            ariaLabel={context.t(ResourceKeys.deviceLists.columns.isPnpDevice)}
                        />
                        {context.t(ResourceKeys.deviceLists.columns.isPnpDevice)}
                    </span>}
            </div>
        );
    }

    private readonly renderCellInterfaceInfo = (device: DeviceSummary, context: LocalizationContextInterface) => {
        const listInterfaces = (deviceSummary: DeviceSummary) => {
            return deviceSummary.interfaceIds && deviceSummary.interfaceIds.length > 0 ? deviceSummary.interfaceIds.join('; ') : '';
        };

        return (
            <div className="device-list-cell-container-content">
                {!!device.isPnpDevice && <span className="device-list-cell-item first no-border">{`${context.t(ResourceKeys.deviceLists.columns.interfaces)}: `}<span className="data">{device.interfaceIds && device.interfaceIds.length > 0 ? listInterfaces(device) : context.t(ResourceKeys.deviceLists.noData)}</span></span>}
            </div>
        );
    }

    private readonly renderLoadingInfo = (context: LocalizationContextInterface) => {
        return (
            <div className="device-list-cell-container-content">
                <span className="device-list-cell-item first">{`${context.t(ResourceKeys.common.loading)}: `}</span>
            </div>
        );
    }

}
