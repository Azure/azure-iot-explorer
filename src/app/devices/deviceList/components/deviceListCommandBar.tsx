/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { CommandBar } from '@fluentui/react';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { ArrayOperation, REFRESH } from '../../../constants/iconNames';

export interface DeviceListCommandBarDataProps {
    disableAdd?: boolean;
    disableRefresh?: boolean;
    disableDelete?: boolean;
}

export interface DeviceListCommandBarActionProps {
    handleAdd: () => void;
    handleRefresh: () => void;
    handleDelete: () => void;
}

export const DeviceListCommandBar: React.FC<DeviceListCommandBarDataProps & DeviceListCommandBarActionProps> = (props: DeviceListCommandBarDataProps & DeviceListCommandBarActionProps) => {
    const { t } = useTranslation();

    const { disableAdd, disableDelete, disableRefresh, handleAdd, handleDelete, handleRefresh } = props;

    return (
        <CommandBar
            items={[
                {
                    ariaLabel: t(ResourceKeys.deviceLists.commands.add),
                    disabled: disableAdd,
                    iconProps: {
                        iconName: ArrayOperation.ADD
                    },
                    key: ArrayOperation.ADD,
                    name: t(ResourceKeys.deviceLists.commands.add),
                    onClick: handleAdd
                },
                {
                    ariaLabel: t(ResourceKeys.deviceLists.commands.refresh),
                    disabled: disableRefresh,
                    iconProps: {
                        iconName: REFRESH
                    },
                    key: REFRESH,
                    name: t(ResourceKeys.deviceLists.commands.refresh),
                    onClick: handleRefresh
                },
                {
                    ariaLabel: t(ResourceKeys.deviceLists.commands.delete.buttonText),
                    disabled: disableDelete,
                    iconProps: {
                        iconName: ArrayOperation.REMOVE
                    },
                    key: ArrayOperation.REMOVE,
                    name: t(ResourceKeys.deviceLists.commands.delete.buttonText),
                    onClick: handleDelete
                },
            ]}
        />
    );
};
