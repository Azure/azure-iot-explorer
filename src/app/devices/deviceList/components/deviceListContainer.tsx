/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import DeviceListComponent, { DeviceListDispatchProps, DeviceListDataProps } from './deviceList';
import { StateType } from '../../../shared/redux/state';
import { listDevicesAction, deleteDevicesAction } from '../actions';
import { getDeviceSummaryListStatus, deviceQuerySelector, deviceSummaryListWrapperNoPNPSelector } from '../selectors';
import DeviceQuery from '../../../api/models/deviceQuery';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';

const mapStateToProps = (state: StateType): DeviceListDataProps => {
    const deviceListSyncStatus = getDeviceSummaryListStatus(state);
    const isFetching = deviceListSyncStatus && deviceListSyncStatus === SynchronizationStatus.working ||  deviceListSyncStatus === SynchronizationStatus.updating;
    return {
        devices: deviceSummaryListWrapperNoPNPSelector(state),
        isFetching,
        query: deviceQuerySelector(state)
    };
};

const mapDispatchToProps = (dispatch: Dispatch): DeviceListDispatchProps => {
    return {
        deleteDevices: (deviceIds: string[]) => dispatch(deleteDevicesAction.started(deviceIds)),
        listDevices: (query?: DeviceQuery) => dispatch(listDevicesAction.started(query)),
    };
};

export default connect(mapStateToProps, mapDispatchToProps, undefined, {
    pure: true,
})(DeviceListComponent);
