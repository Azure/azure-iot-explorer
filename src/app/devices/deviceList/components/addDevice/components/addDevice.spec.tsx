/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { Toggle } from 'office-ui-fabric-react/lib/Toggle';
import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox';
import { CommandBar } from 'office-ui-fabric-react/lib/CommandBar';
import AddDeviceComponent, { AddDeviceDataProps, AddDeviceActionProps, AddDeviceState } from './addDevice';
import { testSnapshot, mountWithLocalization } from '../../../../../shared/utils/testHelpers';
import { MaskedCopyableTextField } from '../../../../../shared/components/maskedCopyableTextField';
import { SynchronizationStatus } from '../../../../../api/models/synchronizationStatus';
import { DeviceAuthenticationType } from '../../../../../api/models/deviceAuthenticationType';
import { DeviceStatus } from '../../../../../api/models/deviceStatus';

describe('components/devices/addDevice', () => {

    const pathname = '/devices/add';
    const location: any = { // tslint:disable-line:no-any
        pathname,
    };
    const routerprops: any = { // tslint:disable-line:no-any
        history: {
            location,
        },
        location,
        match: {
            params: {
            }
        }
    };

    const addDeviceDataProps: AddDeviceDataProps = {
        deviceListSyncStatus: SynchronizationStatus.fetched
    };

    const mockSaveDevice = jest.fn();
    const addDeviceDispatchProps: AddDeviceActionProps = {
        handleSave: mockSaveDevice
    };

    const getComponent = (overrides = {}) => {
        const props = {
            ...addDeviceDataProps,
            ...addDeviceDispatchProps,
            ...routerprops,
            ...overrides
        };

        return (
            <AddDeviceComponent {...props} />
        );
    };

    it('matches snapshot', () => {
        testSnapshot(getComponent());
    });

    it('renders symmetric key input field if not auto generating keys', () => {
        const component = getComponent();
        const wrapper = mountWithLocalization(component, true, true);

        let addDevice = wrapper.find(AddDeviceComponent);
        addDevice.setState({autoGenerateKeys : false});
        wrapper.update();
        expect(wrapper.find(MaskedCopyableTextField).length).toEqual(3); // tslint:disable-line:no-magic-numbers

        wrapper.find(MaskedCopyableTextField).at(1).props().onTextChange('test-key1');
        wrapper.find(MaskedCopyableTextField).at(2).props().onTextChange('test-key2'); // tslint:disable-line:no-magic-numbers
        wrapper.update();
        addDevice = wrapper.find(AddDeviceComponent);
        let addDeviceState = addDevice.state() as AddDeviceState;
        expect(addDeviceState.primaryKey).toEqual('test-key1');
        expect(addDeviceState.primaryKeyError).toEqual('deviceIdentity.validation.invalidKey');
        expect(addDeviceState.secondaryKey).toEqual('test-key2');
        expect(addDeviceState.secondaryKeyError).toEqual('deviceIdentity.validation.invalidKey');

        // auto generate
        const checkbox = addDevice.find(Checkbox);
        checkbox.instance().props.onChange({ target: null}, true);
        addDevice = wrapper.find(AddDeviceComponent);
        addDeviceState = addDevice.state() as AddDeviceState;
        expect(addDeviceState.primaryKey).toEqual(undefined);
        expect(addDeviceState.secondaryKey).toEqual(undefined);
    });

    it('renders selfSigned input field', () => {
        const component = getComponent();
        const wrapper = mountWithLocalization(component, true, true);

        let addDevice = wrapper.find(AddDeviceComponent);
        addDevice.setState({authenticationType : DeviceAuthenticationType.SelfSigned});
        wrapper.update();
        expect(wrapper.find(MaskedCopyableTextField).length).toEqual(3);  // tslint:disable-line:no-magic-numbers

        wrapper.find(MaskedCopyableTextField).at(1).props().onTextChange('test-thumbprint1');
        wrapper.find(MaskedCopyableTextField).at(2).props().onTextChange('test-thumbprint2'); // tslint:disable-line:no-magic-numbers
        wrapper.update();
        addDevice = wrapper.find(AddDeviceComponent);
        const addDeviceState = addDevice.state() as AddDeviceState;
        expect(addDeviceState.primaryThumbprint).toEqual('test-thumbprint1');
        expect(addDeviceState.primaryThumbprintError).toEqual('deviceIdentity.validation.invalidThumbprint');
        expect(addDeviceState.secondaryThumbprint).toEqual('test-thumbprint2');
        expect(addDeviceState.secondaryThumbprintError).toEqual('deviceIdentity.validation.invalidThumbprint');
    });

    it('saves new device identity', () => {
        const component = getComponent();
        const wrapper = mountWithLocalization(component, true, true);

        wrapper.find(MaskedCopyableTextField).first().props().onTextChange('test-device');
        wrapper.find(Toggle).first().instance().props.onChange({ target: null}, false);
        wrapper.update();
        const addDevice = wrapper.find(AddDeviceComponent);
        const addDeviceState = addDevice.state() as AddDeviceState;
        expect(addDeviceState.deviceId).toEqual('test-device');
        expect(addDeviceState.status).toEqual(DeviceStatus.Disabled);
        const commandBar = wrapper.find(CommandBar).first();
        const saveButton = commandBar.props().items[0];
        expect(saveButton.disabled).toBeFalsy();

        const form = addDevice.find('form');
        form.simulate('submit');
        expect(mockSaveDevice).toBeCalled();
    });
});
