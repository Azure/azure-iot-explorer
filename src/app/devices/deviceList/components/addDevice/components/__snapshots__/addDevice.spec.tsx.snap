// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/devices/addDevice matches snapshot 1`] = `
<form
  className="view add-device"
  onSubmit={[Function]}
>
  <div
    className="view-command"
  >
    <StyledCommandBarBase
      items={
        Array [
          Object {
            "ariaLabel": "deviceLists.commands.save",
            "disabled": true,
            "iconProps": Object {
              "iconName": "Save",
            },
            "key": "Save",
            "name": "deviceLists.commands.save",
            "type": "submit",
          },
          Object {
            "ariaLabel": "deviceLists.commands.close",
            "disabled": false,
            "iconProps": Object {
              "iconName": "Cancel",
            },
            "key": "Cancel",
            "name": "deviceLists.commands.close",
            "onClick": [Function],
          },
        ]
      }
    />
  </div>
  <div
    className="view-content view-scroll-vertical"
  >
    <div
      className="form"
    >
      <Connect(MaskedCopyableTextField)
        allowMask={false}
        ariaLabel="deviceIdentity.deviceID"
        error=""
        label="deviceIdentity.deviceID"
        labelCallout="deviceIdentity.deviceIDTooltip"
        onTextChange={[Function]}
        readOnly={false}
        required={true}
        setFocus={true}
        value=""
      />
      <div
        className="authentication"
      >
        <StyledChoiceGroupBase
          label="deviceIdentity.authenticationType.text"
          onChange={[Function]}
          options={
            Array [
              Object {
                "key": "sas",
                "text": "deviceIdentity.authenticationType.symmetricKey.type",
              },
              Object {
                "key": "selfSigned",
                "text": "deviceIdentity.authenticationType.selfSigned.type",
              },
              Object {
                "key": "certificateAuthority",
                "text": "deviceIdentity.authenticationType.ca.type",
              },
            ]
          }
          required={true}
          selectedKey="sas"
        />
        <StyledCheckboxBase
          checked={true}
          className="autoGenerateButton"
          label="deviceIdentity.authenticationType.symmetricKey.autoGenerate"
          onChange={[Function]}
        />
      </div>
      <div
        className="connectivity"
      >
        <Component
          tooltipText="deviceIdentity.hubConnectivity.tooltip"
        >
          deviceIdentity.hubConnectivity.label
        </Component>
        <StyledToggleBase
          ariaLabel="deviceIdentity.hubConnectivity.label"
          checked={true}
          offText="deviceIdentity.hubConnectivity.disable"
          onChange={[Function]}
          onText="deviceIdentity.hubConnectivity.enable"
        />
      </div>
    </div>
  </div>
</form>
`;
